Topics/Development_Process/Error_Handling_in_ExpressJS.md

```diff
@@ -0,0 +1,184 @@
+* **Robust Logging:** Every erroneous request should be logged appropriately. Graceful error handling should ensure that if a developer receives an error response for a request, they should be able to trace the request within the application's log, with at least one entry corresponding to the request error.
+* **Consistent Error Format:** The response bodies for erroneous requests must have consistent formatting. That is, an error object returned for a request bound for endpoint A should have the same formatting as an error object for a request bound for endpoint B. This is to ensure that generic error handlers can be designed by clients to handle various error scenarios by assuming the structure of the returned error object.
+
+## Error Handling Middleware
+Express defines an interface that can be implemented to create a custom error handling middleware (this will override the primitive default error handler). This middleware function resembles general Express middlewares, although an additional `err` parameter is added to the function header. An example error handling middleware is included below:
+
+***errorHandler.js***
+```
+const  DEFAULT_STATUS_CODE = 500;
+
+const  errorHandler = (err, req, res, next) => {
+
+	const  status_code = err.status_code || DEFAULT_STATUS_CODE;
+
+	// More advanced logging libraries may be used here instead.
+	// This should indicate the location where logging occurs.
+	console.log(err);
+	
+	res.status(status_code).send(
+		{
+			error:  err.name,
+			message:  err.message,
+			trace:  err.stack
+		}
+	);
+	next();
+}
+module.exports = errorHandler;
+```
+So what does this all mean? This middleware module is intended to be called every time an error is thrown in a request-response flow. It extracts information from the thrown error object, such as status code, error name, message and a stack trace and then prepares a response with an appropriate status code and an informative body. We will show how this data can be passed to the middleware in the [*Custom Error Objects*](#custom-err-objects) section. The middleware function then calls `next()` to ensure that the next express middleware (if one exists) is called.
+
+Now we must ensure that the error handler middleware is used by our Express app.  The example below shows where to include the error handling  middleware in a generic Express server file (i.e. after routers and all other middlewares). The NPM package `express-async-errors` is used to ensure that the middleware handles both syncronous and asyncronous request-response flows (otherwise the handler will only work with syncronous request-response flows). More information about this package can be found [here](https://www.npmjs.com/package/express-async-errors).
```

Comment:
It may be beneficial to outline the dangers of having a handler only work with synchronous request-response flows, for readers who aren't as experienced with this aspect of express.

Reply:
Great point Ewan! I will add some further context to this

Reply:
Thanks.